Remove 'using namespace std' from header files, properly namespace all std symbols

jonesmz opened the pull request:
I wish to use ninja as a library in another application.

I am aware that the ninja codebase does not attempt to offer any kind of stable API, and I'm not looking for Ninja to do so in the future.

This pull request, if merged, would resolve #1613

My motivations for using the ninja code in a library situation include, but are not limited to, wanting to be able to parse the existing dependency database, and also the ninja log file, to drive some application specific behavior that ninja itself wouldn't benefit from.

There are a variety of other situations where supporting ninja in a library format would also make sense.

E.g. consider a situation where a program written in an interpreted language wanted to use ninja to execute some tasks. That application might want to modify the build graph while executing parts of it based on external information.

Right now, Ninja can't be used in that way, but I strongly believe that from both an architectural point of view, as well as a performance standpoint, moving the Ninja codebase toward a model where it's easier to use as a library (even if that's never officially supported!!!) will do nothing but benefit Ninja. Not the least reason of which is the ease of future maintenance of the code.

There are plenty more reasons, but I don't need to belabor the point.

Regardless, the problem that I continually have, is that it's not practical to directly #include headers from ninja into my existing project. I already *have* data types named map, and string, that live inside their own namespaces. But because ninja has "using namespace std", if I try to use "map" or "string" inside my namespace, but without the namespace indicator, I get ambiguity on which "map" or "string" is actually being used.

I know that Ninja uses "using namespace std" in it's code intentionally. I, of course, disagree with this, as have many others on the mailing list and on github.

So this pull request is a compromise.

Each implementation file can continue using, or not using, "using namespace std" as it sees fit, while header files will use "std::" when referring to types from the std namespace. This makes it significantly less difficult for downstream projects (Not only my own project of course, any downstream project at all) to use Ninja's codebase for their projects.

jhasse commented:
Problem: This will result in lots of merge conflict in all open PRs (which there are a lot of unfortunately).

My idea would be to use `std::` in new code only or when touching a line anyway. Of course it will take a lot longer, before you can use the headers then :/

jonesmz commented:
I don't want to sound unreasonable, but I honestly don't really see why merge conflicts with existing pull requests are a problem.

It should be very easy to update any existing pull requests  to work with this change, especially since it's something that can be done with a simple sed script.

Are there specific pull requests that you would like to point out? I am willing to update them, though I don't see why we should need to do so when many of the existing PRs already have (or used to last I checked) merge conflicts anyway.

If not, how can we plan to execute this change in a way you're willing to accept?

Can we identify specific header files that can be done right now, and then change other header files over time?

If you can give me a specific list, I can modify this pull request to only modify the indicated header files.

Note: The majority of files in this PR are only changed to add "using namespace std;" to the top of the cc files.

Some other proposal?

jonesmz commented:
I looked through all of the open PRs, and approximately 50% of them have merge conflicts as is.

The reason for the merge conflicts appears to be primarily due to the age of these PRs.

I propose the Ninja project adopt the following policy:

PRs will be automatically closed if there is no activity (from anyone) for 3 months.

We can go through the current list of PRs and close any that haven't had any activity for that long.

Or alternatively, we can post a comment on each PR that has no activity for that time, to indicate the new policy is being enacted, and provide a one-time only grace period of 3 months before automatically being closed.

I also suggest the same policy be applied to the list of open issues, with nearly 200 open issues, it's a lot of noise to sift through.

jhasse commented:
> Are there specific pull requests that you would like to point out?

#1600 for example.

> I am willing to update them, though I don't see why we should need to do so when many of the existing PRs already have (or used to last I checked) merge conflicts anyway.

Especially if they already have merge conflicts, we shouldn't make the situation worse!

> Can we identify specific header files that can be done right now, and then change other header files over time?

I don't know. If you want to help, feel free to review some of the open PRs (or if they have merge conflicts, just ask the creator to rebase).

> PRs will be automatically closed if there is no activity (from anyone) for 3 months.

Not a good idea IMHO. I have way too less time to look at most PRs, which doesn't mean it isn't planned to merge them at some point.

For issues, I would also like to keep them open as long as they aren't fixed (or WONTFIX).

jhasse commented:
Replying here so that there isn't too much offtopic in #1600.

> Since this PR is **one of** the main justifications for delaying this PR #1614, it sounds like there is no longer **any** reason to wait to merge #1614.

See emphasis ;)
#1600 was just one example.

jonesmz commented:
Help me to help you.

I asked for specific PRs you consider to be dependencies of this PR.

In the one PR you responded with, you said that it won't be merged any time soon, making any delay caused by the (relatively trivial) merge conflicts that would result into a non-issue.

If you have a specific list, give it to me, and I will communicate with the authors of those PRs if they have an objection to rebasing their PR after this is merged.

jhasse commented:
I don't have a specific list and I don't have the time to create one, sorry. In general I'm trying to fix/merge these next: https://github.com/ninja-build/ninja/milestones/1.10.0

jonesmz commented:
Continued:

Btw, I do apologize for my poor wording.

Please understand that I want to get things moving.

This PR represents a major hassle that I have to deal with every time Ninja merges code. I have to rebase this change every single time.

I want to stop dealing with that.

Coincidentally, this change also improves efforts toward #1499.

Can I assume that this PR will be merged when the list here is merged? https://github.com/ninja-build/ninja/milestones/1.10.0

jhasse commented on src/test.h:
```diff
@@ -144,41 +144,41 @@ struct VirtualFileSystem : public DiskInterface {
+  virtual TimeStamp Stat(const std::string& path, std::string* err) const;
+  virtual bool WriteFile(const std::string& path, const std::string& contents);
+  virtual bool MakeDir(const std::string& path);
+  virtual Status ReadFile(const std::string& path, std::string* contents, std::string* err);
```
Can you clang-format this line (and the others with the same issue) so that it stays in the 80 columns limit?

jonesmz replied:
Yes. I will do so.

jonesmz replied:
Clang formatting the code (using the supplied .clang-format file) results in *lots* of changes.

I suppose I'll edit things by hand then.

jonesmz replied:
Actually.

Can we leave formatting changes, such as the 80 column limit, to a separate PR?

jhasse replied:
Use Visual Studio Code (with C/C++ or ClangFormat extension). You can open the diff of each file and select the changed lines (on the right). Right-click on the selection and select "Format selection". It will only format the selected line then.

You can also add rulers so that you know where you're over the 80 limit to only format those.

jonesmz replied:
I suspect you were writing your reply while I asked:

Can we leave formatting changes, such as the 80 column limit, to a separate PR?

jonesmz replied:
I suggest that .clang-format be run on the whole codebase after https://github.com/ninja-build/ninja/milestones/1.10.0

jhasse replied:
No, sorry. I would like to avoid line touches just for formatting.

jonesmz replied:
You understand that the diff between what we have, and what clang-format gives us is over 5,000 lines right?

clang-format -i src/*.h src/*.cc ; git diff | grep -e "+" -e "-" |  wc -l
8808

You need to rip the bandaid off eventually.

It would be better for *everyone* if we just applied clang format, and moved on.

jhasse replied:
I disagree.

jonesmz replied:
The second commit on this PR addresses lines changed by the first commit that now exceed 80 columns.

jonesmz replied:
> I disagree.

Can you please provide explanations in the future? No meaningful discussion can be had if one party provides explanations for their position, e.g. The clang formatting rules supplied with the project cause over 5 thousand lines of changes

when the other person just says they disagree and that's all.

jhasse replied:
You didn't explain, why it would be better for everyone, if we just applied clang-format and moved on, either. That's why I felt the need to state, that I disagree with that opinion.

jonesmz replied:
Ok, I can see how you thought that.

I thought it was implied by "There are over 5 thousand lines of changes", and the work that contributors have to do by hand in order to comply with the not-in-clang-format rules that the project has.

E.g., it took me 20+ minutes to go line by line and correct the overly long lines in a diff viewer.

Why bother to offer a clang format ruleset if it's not used?

If we apply the rules to the entire repository, and then consistently enforce the use of clang format for any PR that's merged (either by requiring the PR be formatted with clang format prior to merging, or applying clang format immediately after merging), the pain of requiring each contributor to comply individually is condensed into a single "flag day", and then goes away after that.

jhasse replied:
Would have took you only a few minutes if you done it with Visual Studio Code.

> Why bother to offer a clang format ruleset if it's not used?

There's not clang-format ruleset. If in doubt -> use clang-format.

jonesmz replied:
> Would have took you only a few minutes if you done it with Visual Studio Code.

This would require:
1) I have Visual Studio Code, or am able to install it. You don't know what operating system I use, or if it's practical for me to download and install an arbitrary program.
2) That Visual Studio Code is, in fact, faster to use than whatever tool I'm already familiar with.

> There's not clang-format ruleset. If in doubt -> use clang-format.

.clang-format == "clang-format ruleset"

jhasse replied:
> That Visual Studio Code is, in fact, faster to use than whatever tool I'm already familiar with.

Apparently it is, otherwise it wouldn't have took you 20 minutes.

jonesmz replied:
I did a lot more than just re-flow 80+column lines. I compared the clang formatting of every line in the header files to make sure it all looked right.

Visual studio code wouldn't have helped much with that.

jhasse replied:
Ah I see. No need to do that :)

jhasse commented:
> Can I assume that this PR will be merged when the list here is merged? https://github.com/ninja-build/ninja/milestones/1.10.0

Yes, that sounds reasonable.

jonesmz commented:
Anything else this PR needs before merging?

jhasse commented:
I (or someone else) needs to review it.

jonesmz commented:
Anything I'm able to do to accelerate this?

jhasse commented:
Fix #1724 for example.

jonesmz commented:
What does that have to do with a mechanical, code only change?

jhasse commented:
I will fix #1724 prior to this PR. If someone else fixes #1724 before me, I'll have time to look at this PR.

jonesmz commented:
Thank you for clarifying.

jonesmz commented:
Its now been an additional 3 months.

This is a mechanical only change that has no realistic chance of causing problems.

Can we please just merge it? Every time another PR is merged, I have to re-rebase this commit and all of my other patches that depend on it (both patches submitted here as PRs as well as patches elsewhere)

jonesmz commented:
Another month, and now I've had to rebase the changes to fix merge conflicts twice.

Can we please just merge this PR? It changes no functionality. There's nothing to review!

jhasse commented:
I'm sorry and feel your frustration. But #1724 still hasn't been fixed and I could definitely use some help there.

jonesmz commented:
> But #1724 still hasn't been fixed and I could definitely use some help there.

I'm not sympathetic to that bug, so I have no interest in contributing to fixing it.

This PR trivial to review, makes no changes in runtime functionality, and not having it merged is making my life more difficult.

Please just merge it.

jonesmz commented:
Rebased again to account for header files that were modified since this PR was created that were being missed until now.

Additionally: I split out the main commit (not the 80-column commit) into 2.

1. Add 'using namespace std;' to .cc files.
2. Remove 'using namespace std;' from .h files, and fix the resulting compiler errors.

This should make the PR easier to review.

jonesmz commented:
Another month

At the very least merge the first commit. Its fucking tiny.

This entire PR is tiny.

jonesmz commented:
It's been another month, and now #1724 is merged / resolved. Aside from a single comment being added, there were no merge conflicts with that change, so why did this PR need to wait for that one?

Can you please merge this now?

jhasse commented:
I have explained that in another PR: Because that was the issue that was on the top of my todo list and I rarely have enough time to put in for other stuff.

Currently I need to do the 1.10.1 release next. Please stop bugging me.

jonesmz commented:
This is my last message about this PR, and potentially my last interaction with Ninja as a project entirely.

I can't believe that you've seriously let such a trivial PR sit for over a year. It's degrading to contributors and seriously diminishes the credibility of the Ninja project as code that can be relied on.

Your inability to handle the basic maintenance tasks for this project in a timely manner have serious ramifications for contributors across the world. I personally experienced not one, but two, projects I was involved in at work get cancelled *specifically* because of this project's maintainers inability to handle trivial changes.

If the maintainers of Ninja are unable to do simple things in a timely manner, find someone who can.

jhasse commented:
It's hard to find good people for that salary.

jhasse commented on src/build.h:
```diff
@@ -257,7 +257,7 @@ struct BuildStatus {
   /// @param status The status of the edge.
-  string FormatProgressStatus(const char* progress_status_format,
+  std::string FormatProgressStatus(const char* progress_status_format,
                               EdgeStatus status) const;
```
fix alignment

jhasse commented on src/clparser.h:
```diff
@@ -27,26 +26,26 @@ struct CLParser {
-  static string FilterShowIncludes(const string& line,
-                                   const string& deps_prefix);
+  static std::string FilterShowIncludes(const std::string& line,
+                                   const std::string& deps_prefix);
```
alignment

jhasse commented on src/eval_env.h:
```diff
@@ -27,18 +26,18 @@ struct Rule;
 };
 
-/// A tokenized string that contains variable references.
+/// A tokenized std::string that contains variable references.
```
I think this was meant more as a general string, not a std::string in particular.

jhasse commented on src/eval_env.h:
```diff
@@ -84,26 +83,26 @@ struct BindingEnv : public Env {
   /// This function takes as parameters the necessary info to do (2).
-  string LookupWithFallback(const string& var, const EvalString* eval,
+  std::string LookupWithFallback(const std::string& var, const EvalString* eval,
                             Env* env);
```
alignment

jhasse commented on src/graph.h:
```diff
@@ -74,12 +73,12 @@ struct Node {
   }
-  static string PathDecanonicalized(const string& path,
+  static std::string PathDecanonicalized(const std::string& path,
                                     uint64_t slash_bits);
```
alignment

jhasse commented on src/includes_normalize.h:
```diff
@@ -14,26 +14,27 @@
-  static string Relativize(StringPiece path,
-                           const vector<StringPiece>& start_list, string* err);
+  static std::string AbsPath(StringPiece s, std::string* err);
+  static std::string Relativize(StringPiece                     path,
```
Did clang-format do this? If not, please don't.

jhasse commented on src/manifest_parser.h:
```diff
@@ -44,24 +44,25 @@ struct ManifestParser : public Parser {
 private:
-  /// Parse a file, given its contents as a string.
-  bool Parse(const string& filename, const string& input, string* err);
+  /// Parse a file, given its contents as a std::string.
```
remove std:: here

jhasse commented on src/string_piece.h:
```diff
@@ -17,11 +17,9 @@
 #include <string.h>
 
-/// StringPiece represents a slice of a string whose memory is managed
+/// StringPiece represents a slice of a std::string whose memory is managed
```
also don't mention std:: here

jhasse commented on src/string_piece.h:
```diff
@@ -30,22 +28,23 @@ struct StringPiece {
-  /// data into a new string.
-  string AsString() const {
-    return len_ ? string(str_, len_) : string();
+  /// Convert the slice into a full-fledged std::std::string, copying the
```
You've added two std:: here.

jhasse commented on src/string_piece.h:
```diff
@@ -30,22 +28,23 @@ struct StringPiece {
-  string AsString() const {
-    return len_ ? string(str_, len_) : string();
+  /// Convert the slice into a full-fledged std::std::string, copying the
+  /// data into a new std::string.
```
keep string without std::

jhasse reviewed (changes requested):
Have you checked if we can enable https://clang.llvm.org/extra/clang-tidy/checks/google-global-names-in-headers.html now?
